python/tests/test_combinatorics.py

```diff
@@ -420,3 +445,516 @@ def test_is_symmetrical(self):
+
+        sample_sets = [[0], [-1]]
+        self.verify_node_out_of_bounds_error(tables.tree_sequence(), sample_sets)
+
+    def verify_value_error(self, ts, sample_sets=None):
+        with self.assertRaises(ValueError):
+            ts.first().count_topologies(sample_sets)
+        with self.assertRaises(ValueError):
+            list(ts.count_topologies(sample_sets))
+
+    def verify_node_out_of_bounds_error(self, ts, sample_sets=None):
+        with self.assertRaises(ValueError):
+            ts.first().count_topologies(sample_sets)
+        with self.assertRaises(IndexError):
+            list(ts.count_topologies(sample_sets))
+
+    def test_standard_msprime_migrations(self):
+        for num_populations in range(2, 5):
+            samples = [5] * num_populations
+            ts = self.simulate_multiple_populations(samples)
+            self.verify_topologies(ts)
+
+    def simulate_multiple_populations(self, sample_sizes):
+        d = len(sample_sizes)
+        M = 0.2
+        m = M / (2 * (d - 1))
+
+        migration_matrix = [
+            [m if k < d and k == i + 1 else 0 for k in range(d)] for i in range(d)
+        ]
+
+        pop_configurations = [
+            msprime.PopulationConfiguration(sample_size=size) for size in sample_sizes
+        ]
+        return msprime.simulate(
+            population_configurations=pop_configurations,
+            migration_matrix=migration_matrix,
+            recombination_rate=0.1,
+        )
+
+    def test_msprime_dtwf(self):
+        migration_matrix = np.zeros((4, 4))
+        population_configurations = [
+            msprime.PopulationConfiguration(
+                sample_size=10, initial_size=10, growth_rate=0
+            ),
+            msprime.PopulationConfiguration(
+                sample_size=10, initial_size=10, growth_rate=0
+            ),
+            msprime.PopulationConfiguration(
+                sample_size=10, initial_size=10, growth_rate=0
+            ),
+            msprime.PopulationConfiguration(
+                sample_size=0, initial_size=10, growth_rate=0
+            ),
+        ]
+        demographic_events = [
+            msprime.PopulationParametersChange(population=1, time=0.1, initial_size=5),
+            msprime.PopulationParametersChange(population=0, time=0.2, initial_size=5),
+            msprime.MassMigration(time=1.1, source=0, dest=2),
+            msprime.MassMigration(time=1.2, source=1, dest=3),
+            msprime.MigrationRateChange(time=2.1, rate=0.3, matrix_index=(2, 3)),
+            msprime.MigrationRateChange(time=2.2, rate=0.3, matrix_index=(3, 2)),
+        ]
+        ts = msprime.simulate(
+            migration_matrix=migration_matrix,
+            population_configurations=population_configurations,
+            demographic_events=demographic_events,
+            random_seed=2,
+            model="dtwf",
+        )
+
+        self.verify_topologies(ts)
+
+    def test_forward_time_wright_fisher(self):
```

Comment:
Be a little nicer to split this up into three tests, which are named. That way we can see the different types of forward sim topology that we're testing against. Big monolithic tests are hard to parse quicky for the reader.